diff --git a/src/sprintf.js b/src/sprintf.js
index 65d6324645ef1d0133614a71497751009dc5c912..c6b88711fa8da88bc879955bd7d5bfe6343d6cc1 100644
--- a/src/sprintf.js
+++ b/src/sprintf.js
@@ -66,6 +66,7 @@
                     is_positive = arg >= 0
                 }
 
+                var precision = ph.precision === undefined ? undefined : Math.min(100, Math.max(ph.type === 'g' ? 1 : 0, Number(ph.precision)))
                 switch (ph.type) {
                     case 'b':
                         arg = parseInt(arg, 10).toString(2)
@@ -81,13 +82,13 @@
                         arg = JSON.stringify(arg, null, ph.width ? parseInt(ph.width) : 0)
                         break
                     case 'e':
-                        arg = ph.precision ? parseFloat(arg).toExponential(ph.precision) : parseFloat(arg).toExponential()
+                        arg = ph.precision ? parseFloat(arg).toExponential(precision) : parseFloat(arg).toExponential()
                         break
                     case 'f':
-                        arg = ph.precision ? parseFloat(arg).toFixed(ph.precision) : parseFloat(arg)
+                        arg = ph.precision ? parseFloat(arg).toFixed(precision) : parseFloat(arg)
                         break
                     case 'g':
-                        arg = ph.precision ? String(Number(arg.toPrecision(ph.precision))) : parseFloat(arg)
+                        arg = ph.precision ? String(Number(arg.toPrecision(precision))) : parseFloat(arg)
                         break
                     case 'o':
                         arg = (parseInt(arg, 10) >>> 0).toString(8)
